In [35]:

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasRegressor

In [36]:
pip install tensorflow

Note: you may need to restart the kernel to use updated packages.


For this example, use a linear activation function within the keras library to create a regression-based neural network. 
Essentially, we are trying to predict the value of a potential car sale (i.e. how much a particular person will spend on buying a car) for a customer based on the following attributes:

-Age

-Gender

-Average miles driven per day

-Personal debt

-Monthly income

In [49]:
dataset=np.loadtxt("cars.csv", delimiter=",")
x=dataset[:,0:5]#predictor variables(age,gender,miles driven,debt,income)
y=dataset[:,5] #target(potential car sale)
y=np.reshape(y, (-1,1))
scaler = MinMaxScaler()#to normalize the data for neural network
print(scaler.fit(x))
print(scaler.fit(y))
x_scale=scaler.transform(x)
y_scale=scaler.transform(y)

MinMaxScaler(copy=True, feature_range=(0, 1))
MinMaxScaler(copy=True, feature_range=(0, 1))


In [38]:
df=pd.read_csv('cars.csv')

In [50]:
X_train, X_test, y_train, y_test = train_test_split(x_scale, y_scale)

# Create the model.

training the neural network -12 & 8 neurons in the hidden layer with 5 inputs(age,gender,miles driven,debt,income) and linear activation function

In [52]:
model = Sequential()
model.add(Dense(12, input_dim=5, kernel_initializer='normal', activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='linear'))
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_13 (Dense)             (None, 12)                72        
_________________________________________________________________
dense_14 (Dense)             (None, 8)                 104       
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 9         
Total params: 185
Trainable params: 185
Non-trainable params: 0
_________________________________________________________________


# compile/fit the model

mse and mae give the test error which is the accuracy of the model in predicting the data.lower the mse and mae better the accuracy of the model.

In [54]:
model.compile(loss='mse', optimizer='adam', metrics=['mse','mae'])
Train_data= model.fit(X_train, y_train, epochs=150, batch_size=50,  verbose=1, validation_split=0.2)

Train on 577 samples, validate on 145 samples
Epoch 1/150
577/577 [==============================] - 0s 668us/step - loss: 0.0173 - mean_squared_error: 0.0173 - mean_absolute_error: 0.0971 - val_loss: 0.0213 - val_mean_squared_error: 0.0213 - val_mean_absolute_error: 0.1130
Epoch 2/150
577/577 [==============================] - 0s 42us/step - loss: 0.0172 - mean_squared_error: 0.0172 - mean_absolute_error: 0.0968 - val_loss: 0.0213 - val_mean_squared_error: 0.0213 - val_mean_absolute_error: 0.1131
Epoch 3/150
577/577 [==============================] - 0s 42us/step - loss: 0.0172 - mean_squared_error: 0.0172 - mean_absolute_error: 0.0968 - val_loss: 0.0214 - val_mean_squared_error: 0.0214 - val_mean_absolute_error: 0.1130
Epoch 4/150
577/577 [==============================] - 0s 53us/step - loss: 0.0172 - mean_squared_error: 0.0172 - mean_absolute_error: 0.0968 - val_loss: 0.0211 - val_mean_squared_error: 0.0211 - val_mean_absolute_error: 0.1124
Epoch 5/150
577/577 [====================

Epoch 37/150
577/577 [==============================] - 0s 53us/step - loss: 0.0168 - mean_squared_error: 0.0168 - mean_absolute_error: 0.0942 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1115
Epoch 38/150
577/577 [==============================] - 0s 46us/step - loss: 0.0169 - mean_squared_error: 0.0169 - mean_absolute_error: 0.0951 - val_loss: 0.0211 - val_mean_squared_error: 0.0211 - val_mean_absolute_error: 0.1118
Epoch 39/150
577/577 [==============================] - 0s 42us/step - loss: 0.0168 - mean_squared_error: 0.0168 - mean_absolute_error: 0.0941 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1114
Epoch 40/150
577/577 [==============================] - 0s 40us/step - loss: 0.0168 - mean_squared_error: 0.0168 - mean_absolute_error: 0.0943 - val_loss: 0.0211 - val_mean_squared_error: 0.0211 - val_mean_absolute_error: 0.1117
Epoch 41/150
577/577 [==============================] - 0s 45us/step - loss: 0.0168 

Epoch 73/150
577/577 [==============================] - 0s 42us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0935 - val_loss: 0.0208 - val_mean_squared_error: 0.0208 - val_mean_absolute_error: 0.1106
Epoch 74/150
577/577 [==============================] - 0s 55us/step - loss: 0.0168 - mean_squared_error: 0.0168 - mean_absolute_error: 0.0934 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1111
Epoch 75/150
577/577 [==============================] - 0s 42us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0941 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1110
Epoch 76/150
577/577 [==============================] - 0s 39us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0933 - val_loss: 0.0208 - val_mean_squared_error: 0.0208 - val_mean_absolute_error: 0.1106
Epoch 77/150
577/577 [==============================] - 0s 37us/step - loss: 0.0167 

Epoch 109/150
577/577 [==============================] - 0s 40us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0929 - val_loss: 0.0209 - val_mean_squared_error: 0.0209 - val_mean_absolute_error: 0.1107
Epoch 110/150
577/577 [==============================] - 0s 46us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0937 - val_loss: 0.0208 - val_mean_squared_error: 0.0208 - val_mean_absolute_error: 0.1104
Epoch 111/150
577/577 [==============================] - 0s 47us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0930 - val_loss: 0.0209 - val_mean_squared_error: 0.0209 - val_mean_absolute_error: 0.1105
Epoch 112/150
577/577 [==============================] - 0s 43us/step - loss: 0.0166 - mean_squared_error: 0.0166 - mean_absolute_error: 0.0930 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1108
Epoch 113/150
577/577 [==============================] - 0s 37us/step - loss: 0.

Epoch 145/150
577/577 [==============================] - 0s 40us/step - loss: 0.0166 - mean_squared_error: 0.0166 - mean_absolute_error: 0.0930 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1108
Epoch 146/150
577/577 [==============================] - 0s 35us/step - loss: 0.0167 - mean_squared_error: 0.0167 - mean_absolute_error: 0.0928 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1109
Epoch 147/150
577/577 [==============================] - 0s 43us/step - loss: 0.0166 - mean_squared_error: 0.0166 - mean_absolute_error: 0.0930 - val_loss: 0.0210 - val_mean_squared_error: 0.0210 - val_mean_absolute_error: 0.1108
Epoch 148/150
577/577 [==============================] - 0s 38us/step - loss: 0.0166 - mean_squared_error: 0.0166 - mean_absolute_error: 0.0927 - val_loss: 0.0209 - val_mean_squared_error: 0.0209 - val_mean_absolute_error: 0.1105
Epoch 149/150
577/577 [==============================] - 0s 40us/step - loss: 0.

In [42]:
Xnew = np.array([[40, 0, 26, 9000, 8000]])

# predict the new set of data

In [55]:
Xnew = np.array([[40, 0, 26, 9000, 8000]])
Xnew= scaler.transform(Xnew)
ynew= model.predict(Xnew)

ynew = scaler.inverse_transform(ynew) 
Xnew = scaler.inverse_transform(Xnew)
Xnew[0], ynew

(array([  40.,    0.,   26., 9000., 8000.]),
 array([[12833.113]], dtype=float32))

In [56]:
ynew

array([[12833.113]], dtype=float32)

For the given inputs of age-40,female,miles driven 26 with debt of 9000 and income of 8000 the potential car sale is 12833.113 as predicted by the model.